release/src/router/plc-utils/nda/plcmcinfo.1
.TH plcmcinfo 1 "Mar 2014" "plc-utils-2.1.6" "Qualcomm Atheros Powerline Toolkit"

.SH NAME
plcmcinfo - Qualcomm Atheros Panther/Lynx Powerline Device Manager

.SH SYNOPSIS
.BR plcmcinfo
.RI [ options ] 
.RI [ device ] 
.RI [ device ] 
[...]

.SH DESCRIPTION

.PP
Read and display multicast group information returned by the Multicast Group Information message (VS_MULTICAST_INFO).
The message fetch all, or selected, multicast group information learned by the PLC device from IGMP and MLD snooping.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBplc\fR man page for an overview and installation instructions.

.SH OPTIONS

.TP
-\fBa \fIrequest_address_type\fR
The request address type where \fB0\fR=\fBIPv4\fR, \fB1\fR=\fBIPv6\fR and \fB2\fR=\fBMAC\fR.
The request address type may be specified in decimal, hexadecimal or binary or with one of the case sensitive strings "\fBIPv4\fR", "\fBIPv6\fR" or "\fBMAC\fR".
Prefix a hexadecimal value with "\fB0x\fR".
Prefix a binary value with "\fB0b\fR".
This option accepted but is not supported on some program versions.

.TP
-\fBi\fR \fIinterface\fR
Select the host Ethernet interface.
All requests are sent via this host interface and only reponses received via this host interface are recognized.
The default interface is \fBeth1\fR because most people use \fBeth0\fR as their principle network connection; however, if environment string "PLC" is defined then it takes precedence over the default interface.
This option then takes precedence over either default.

.TP
.RB - q
Suppresses status messages on stderr.

.TP
-\fBt \fIrequest_type\fR
The action take during the request such as the number or type of groups to include, whether to clear or overwrite data and the type of snooping to perform.
The request type may be specified as a decimal, hexadecimal or binary.
Prefix a hexadecimal value with "\fB0x\fR".
Prefix a binary value with "\fB0b\fR".
This option accepted but is not supported on some program versions.

.TP
.RB - v
Print additional information on stdout.
In particular, this option dumps incoming and outgoing packets which can be saved as text files for reference.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over other options on the command line.

.TP
.RB - ? ,-- version
Print program version information on stdout.
This option takes precedence over other options on the command line.
Use this option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.

.SH ARGUMENTS

.TP
.IR device
The Ethernet hardware address of some powerline device.
More than one address may be specified on the command line.
If more than one address is specified then operations are performed on each device in turn.
The default address is \fBlocal\fR.
as explained in the \fBDEVICES\fR section.

.SH KEYS
Passwords are variable length character strings that end-users can remember.
Keys are fixed length binary values created by encrypting passwords.
There are two encryption algorithms for HomePlugAV.
One for DAKs and the other for NMKs.
This means that a given password will produce different keys depending on use.
This program only deals with keys because that is what powerline devices recognize.
The passwords that generated the keys are irrelevant here.

.PP
Encryption keys are tedious to type and prone to error.
For convenience, symbolic names have been assigned to common encryption keys and are recognized by options -\fBD\fR and -\fBK\fR.

.TP
.BR key1
Key for encrypted password "\fBHomePlugAV\fR".
This is "689F074B8B0275A2710B0B5779AD1630" for option -\fBD\fR and "50D3E4933F855B7040784DF815AA8DB7" for option -\fBK\fR.

.TP
.BR key2
Key for encrypted password "\fBHomePlugAV0123\fR".
This is "F084B4E8F6069FF1300C9BDB812367FF" for option -\fBD\fR and "B59319D7E8157BA001B018669CCEE30D" for option -\fBK\fR.

.TP
.BR none
Always "00000000000000000000000000000000".

.SH DEVICES
Powerline devices use Ethernet hardware, or Media Access Control (MAC), addresses.
Device addresses are 12 hexadecimal digits (\fB0123456789ABCDEFabcdef\fR) in upper, lower or mixed case.
Individual octets may be separated by colons, for clarity, but not all octets need to be seperated.
For example, "00b052000001", "00:b0:52:00:00:01" and "00b052:000001" are valid and equivalent.

.PP
These symbolic addresses are recognized by this program and may be used instead of the actual address value.

.TP
.BR all
Equivalent to "broadcast", described next.

.TP
.BR broadcast
A synonym for the standard Ethernet broadcast address, \fBFF:FF:FF:FF:FF:FF\fR.
All devices, whether local, remote or foreign will respond to this address.

.TP
.BR local
A synonym for the Qualcomm Atheros vendor specific Local Management Address (LMA), \fB00:B0:52:00:00:01\fR.
All local Atheros devices will recognize this address but remote and foreign devices will not.
A remote device is any device at the far end of a powerline connection.
A foreign device is any device not manufactured by Atheros.

.SH REFERENCES
See the Qualcomm Atheros HomePlug AV Firmware Technical Reference Manual for more information.

.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information may not be available.
Qualcomm Atheros reserves the right to modify message structure or content in future firmware releases without any obligation to notify or compensate users of this program.

.SH EXAMPLES
None.

.SH SEE ALSO
.BR plc ( 1 ),
.BR plcrule ( 1 ),
.BR plctool ( 1 )

.SH CREDITS
 Charles Maier <cmaier@qca.qualcomm.com>



release/src/router/plc-utils/nda/plcmcinfo.c
/*====================================================================*
*
*   Copyright (c) 2013 Qualcomm Atheros, Inc.
*
*   All rights reserved.
*
*====================================================================*/

/*====================================================================*
 *
 *   plcmcinfo.c -
 *
 *   Contributor(s):
 *      Charles Maier <cmaier@qca.qualcomm.com>
 *      Nathaniel Houghton <nhoughto@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>
#include <limits.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../ether/channel.h"
#include "../key/HPAVKey.h"
#include "../key/keys.h"
#include "../ram/sdram.h"
#include "../pib/pib.h"
#include "../nvm/nvm.h"
#include "../plc/plc.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../plc/Confirm.c"
#include "../plc/Devices.c"
#include "../plc/Display.c"
#include "../plc/Failure.c"
#include "../plc/ReadFMI.c"
#include "../plc/ReadMME.c"
#include "../plc/SendMME.c"
#include "../plc/Request.c"
#include "../plc/chipset.c"
#endif

#ifndef MAKEFILE
#include "../nda/MulticastInfo1.c"
#include "../nda/MulticastInfo2.c"
#endif

#ifndef MAKEFILE
#include "../tools/error.c"
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/synonym.c"
#include "../tools/uintspec.c"
#include "../tools/basespec.c"
#include "../tools/version.c"
#include "../tools/hexdump.c"
#include "../tools/hexencode.c"
#include "../tools/hexdecode.c"
#include "../tools/hexstring.c"
#include "../tools/decstring.c"
#include "../tools/decdecode.c"
#include "../tools/hexout.c"
#include "../tools/todigit.c"
#include "../tools/checksum32.c"
#include "../tools/strfbits.c"
#include "../tools/typename.c"
#endif

#ifndef MAKEFILE
#include "../ether/channel.c"
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#endif

#ifndef MAKEFILE
#include "../ram/nvram.c"
#include "../ram/sdramfile.c"
#include "../ram/sdrampeek.c"
#endif

#ifndef MAKEFILE
#include "../mme/EthernetHeader.c"
#include "../mme/QualcommHeader.c"
#include "../mme/QualcommHeader1.c"
#include "../mme/UnwantedMessage.c"
#include "../mme/MMECode.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define PLCMCINFO_ADDRESS_TYPE 0
#define PLCMCINFO_REQUEST_TYPE 0

#define PLCMCINFO_CLASSIC (1 << 0)

/*====================================================================*
 *   program variables;
 *--------------------------------------------------------------------*/

static struct _term_ const addresses [] =

{
	{
		"IPv4",
		"0x00"
	},
	{
		"IPv6",
		"0x01"
	},
	{
		"MAC",
		"0x03"
	}
};

/*====================================================================*
 *
 *   int main (int argc, char const * argv[]);
 *
 *   parse command line, populate plc structure and perform selected
 *   operations; show help summary if asked; see getoptv and putoptv
 *   to understand command line parsing and help summary display; see
 *   plc.h for the definition of struct plc;
 *
 *   the command line accepts multiple MAC addresses and the program
 *   performs the specified operations on each address, in turn; the
 *   address order is significant but the option order is not; the
 *   default address is a local broadcast that causes all devices on
 *   the local H1 interface to respond but not those at the remote
 *   end of the powerline;
 *
 *   the default address is 00:B0:52:00:00:01; omitting the address
 *   will automatically address the local device; some options will
 *   cancel themselves if this makes no sense;
 *
 *   the default interface is eth1 because most people use eth0 as
 *   their principle network connection; you can specify another
 *   interface with -i or define environment string PLC to make
 *   that the default interface and save typing;
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	static char const * optv [] =
	{
		"a:cegi:qt:v",
		"device [device] [...]",
		"Qualcomm Atheros Multicast Information Probe",
		"a n\taddress type is (n) [" LITERAL (PLCMCINFO_ADDRESS_TYPE) "]",
		"c\tclassic method",
		"e\tredirect stderr to stdout",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"q\tquiet mode",
		"t n\trequest type is (n) [" LITERAL (PLCMCINFO_REQUEST_TYPE) "]",
		"v\tverbose mode",
		(char const *) (0)
	};

#include "../plc/plc.c"

	flag_t flags = (flag_t) (0);
	signed c;
	plc.action = PLCMCINFO_ADDRESS_TYPE;
	if (getenv (PLCDEVICE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (PLCDEVICE));

#else

		channel.ifname = strdup (getenv (PLCDEVICE));

#endif

	}
	optind = 1;
	while (~ (c = getoptv (argc, argv, optv)))
	{
		switch (c)
		{
		case 'a':
			plc.sector = (unsigned) (uintspec (synonym (optarg, addresses, SIZEOF (addresses)), 0, UCHAR_MAX));
			error (1, ENOTSUP, "Address Type");
			break;
		case 'c':
			_setbits (flags, PLCMCINFO_CLASSIC);
			break;
		case 'e':
			dup2 (STDOUT_FILENO, STDERR_FILENO);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 't':
			plc.action = (unsigned) (uintspec (optarg, 0, UCHAR_MAX));
			error (1, ENOTSUP, "Request Type");
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			_setbits (plc.flags, PLC_VERBOSE);
			break;
		default: 
			break;
		}
	}
	argc -= optind;
	argv += optind;
	openchannel (& channel);
	if (! (plc.message = malloc (sizeof (* plc.message))))
	{
		error (1, errno, PLC_NOMEMORY);
	}
	if (! argc)
	{
		if (_anyset (flags, PLCMCINFO_CLASSIC))
		{
			MulticastInfo1 (& plc);
		}
		else 
		{
			MulticastInfo2 (& plc);
		}
	}
	while ((argc) && (* argv))
	{
		if (! hexencode (channel.peer, sizeof (channel.peer), synonym (* argv, devices, SIZEOF (devices))))
		{
			error (1, errno, PLC_BAD_MAC, * argv);
		}
		if (_anyset (flags, PLCMCINFO_CLASSIC))
		{
			MulticastInfo1 (& plc);
		}
		else 
		{
			MulticastInfo2 (& plc);
		}
		argc--;
		argv++;
	}
	free (plc.message);
	closechannel (& channel);
	exit (0);
}

